### ISSUE TITLE
MasterWeightOptimizer: per-resume precision drift from bf16 master rebuild

### ISSUE THREAD

**Issue by shuheng-liu** [TensorAuto/OpenTau#185]
## Summary

`MasterWeightOptimizer` (introduced in #182) keeps fp32 master copies of bf16 model weights. `accelerator.save_state` persists the model's bf16 weights and the inner optimizer's state (`exp_avg` / `exp_avg_sq`) — but **not** the fp32 master parameter values themselves, because `torch.optim.Optimizer.state_dict` doesn't capture parameter tensors, only optimizer state.

On resume, `accelerator.load_state` restores the bf16 live weights, then `MasterWeightOptimizer.rebuild_masters_from_live` re-derives the fp32 masters from those bf16 weights:

```python
new_data = live.data.detach().to(torch.float32).clone()
```

This loses the fp32 mantissa bits accumulated since the last checkpoint. Bounded — one round-off per resume, not compounding within a single run — but a real divergence from DeepSpeed's `BF16_Optimizer`, which checkpoints the fp32 masters explicitly and avoids this entirely.

## Why it usually doesn't matter

Each resume drops ~16 bits of mantissa from the parameters used to seed the next step. Adam's update reads the master, applies an fp32 step, and the result rejoins the fp32 trajectory. The error is bounded by `eps_bf16 ≈ 2^-7 * |master|` per resume; subsequent training in fp32 quickly washes it out (similar dynamics to the pre-#182 → post-#182 resume tax in #<bug3-issue>).

## Why it might matter

Frequent / long-running checkpoint cadences — e.g., a job that's preempted and resumed nightly for weeks — pay this tax repeatedly. Each individual hit is small; the cumulative effect on a 100-resume training run is unmeasured. Probably below noise but unverified.

## Options

1. **Document and accept** (current state). DDP throughput and memory budget were the dominant constraints; the loss is bounded.
2. **Save fp32 masters explicitly.** Add a custom `state_dict` / `load_state_dict` path that serialises `_fp32_params` alongside the inner Adam state, restoring them verbatim on resume. Cost: roughly doubles the optimizer-state checkpoint size (Adam state is already 2× params; fp32 masters add 1× more). For pi05 (~3.4B params): an extra ~13.6 GB on the checkpoint.
3. **Save fp32 masters compressed.** Save the **delta** between the fp32 master and the bf16 live weights instead of the full fp32 master. The delta lives in the bottom ~16 mantissa bits, so it compresses well; could store as int16. Trades complexity for size.

Option 2 is the obvious right answer if anyone reports an empirical accuracy gap; option 3 is engineering for a hypothetical that may not exist.

## Owner / scope

* Touches: `src/opentau/optim/master_weights.py::state_dict` / `load_state_dict`.
* `train.py` would need to drop the post-`load_state` `rebuild_masters_from_live` call when fp32 masters are restored verbatim (the post-`prepare` device-migration call still stays).
* Test: round-trip a `state_dict` and assert master values are bit-identical.

Refs: #181, #182.
